### ISSUE TITLE
Retrieval of device name slow on NVIDIA GPUs

### ISSUE THREAD

**Issue by CNugteren** [CNugteren/CLBlast#193]
It turns out the cost of retrieving the OpenCL device name (`CL_DEVICE_NAME`) is very high on NVIDIA GPUs. Since CLBlast needs the name to query the database, this can become a performance issue for smaller matrix sizes.

CLBlast now has a diagnostics helper (`clblast_test_diagnostics`) to verify this. Indeed, here is the output for two different NVIDIA GPUs with different versions of CUDA:

```
 --- OpenCL device naming:
* Device type                   GPU
* Device name                   TITAN X (Pascal)
* Platform vendor               NVIDIA Corporation
* Platform version              OpenCL 1.2 CUDA 8.0.0

 --- Some OpenCL library benchmarks (functions from clpp11.h):
* queue.GetContext()            0.0004 ms
* queue.GetDevice()             0.0002 ms
* device.Name()                 0.1651 ms
* Buffer<float>(context, 1024)  0.0021 ms
```
and:
```
 --- OpenCL device naming:
* Device type                   GPU
* Device name                   GeForce GTX 470
* Platform vendor               NVIDIA Corporation
* Platform version              OpenCL 1.1 CUDA 6.0.1

 --- Some OpenCL library benchmarks (functions from clpp11.h):
* queue.GetContext()            0.0000 ms
* queue.GetDevice()             0.0000 ms
* device.Name()                 0.0710 ms
* Buffer<float>(context, 1024)  0.0010 ms
```

Both are significantly slower at retrieving the device name compared to other devices, e.g. an Intel GPU using Beignet:
```
 --- OpenCL device naming:
* Device type                   GPU
* Device name                   Intel(R) HD Graphics Skylake ULT GT2
* Platform vendor               Intel
* Platform version              OpenCL 2.0 beignet 1.3

 --- Some OpenCL library benchmarks (functions from clpp11.h):
* queue.GetContext()            0.0004 ms
* queue.GetDevice()             0.0002 ms
* device.Name()                 0.0003 ms
* Buffer<float>(context, 1024)  0.0038 ms
```

I propose to solve this by caching the device name based on the OpenCL queue and thus paying this overhead only once.

**Comment by blueberry** [comment]
Is it cached after the retrieval and reused for subsequent calls, or it is *per every call*?

**Comment by CNugteren** [comment]
Not yet, I plan to build that into CLBlast. Right now every time the device name is required (which is actually two calls to `clGetDeviceInfo`) is expensive. I just wanted to report this here, because I was quite surprised that this is so expensive. And with e.g. a Titan X doing a m=n=k=1024 in less than 0.5 ms, an overhead of 0.1ms is a lot!

**Comment by blueberry** [comment]
Why is that used at all? Isn't it much easier and more efficient to use the device pointer/object as a key for such identification? At startup/initialization/call we could get the necessary attributes (such as device name), to identify which kernels to invoke, and then use that info in subsequent calls? That is the information that does not change during the lifetime of the application.

I admit that I am not using OpenCL from C, but from Clojure, where everything is much easier and more convenient to keep track of, so maybe I'm missing something. But definitely it seems to me that this kind of management would be easier to do if there was a clear separation of actual kernel setup/enqueuement and the compilation/caching etc.

That's why I asked for that bulk startup initialization procedure last year. Hardware might be different on different systems, but one system only see one or only a few different kinds of hardware....

**Comment by CNugteren** [comment]
Yes, you are right in questioning why the device name is used at all. I started working on this an hour ago and realised the same. It turns out the device name was used for storing a cache of the database as well. I've changed that now such that it is based on the platform ID and device ID. So then the only use of the device name now is when compiling the kernel (to find the database parameters). And that's actually a one-time and very costly thing, so I believe the issue is solved with commit 890281f !

Here are some benchmarks on a relatively old NVIDIA GPU (GTX 470):
```
Before the changes:
        m;        n;        k;       ms_1; GFLOPS_1;
       64;       64;       64;       0.14;      3.9;
      128;      128;      128;       0.15;     28.0;
      192;      192;      192;       0.17;     80.9;
      256;      256;      256;       0.23;    142.8;
      320;      320;      320;       0.33;    196.8;
      384;      384;      384;       0.46;    246.7;
      448;      448;      448;       0.62;    288.7;
      512;      512;      512;       0.88;    306.1;
After the changes:
        m;        n;        k;       ms_1; GFLOPS_1;
       64;       64;       64;       0.05;      9.9;
      128;      128;      128;       0.07;     61.7;
      192;      192;      192;       0.09;    153.9;
      256;      256;      256;       0.15;    219.3;
      320;      320;      320;       0.25;    261.1;
      384;      384;      384;       0.38;    300.4;
      448;      448;      448;       0.54;    333.0;
      512;      512;      512;       0.80;    337.2;
```
Notice the gain of ~0.1ms across the board, which is of course most important for small problems, but turns out to become quite an issue for larger problems as well on modern Pascal GPUs.
